### PR TITLE
[core][experimental] Allocate a channel for each InputAttributeNode

### LINKED ISSUE #47366: [core][experimental] Allocate a channel for each InputAttributeNode
### Description

Currently, InputNode has only one channel for all InputAttributeNodes. There are two questions:

1. We might want to use different kinds of channels for different `InputAttributeNodes`.
2. `InputNode` needs to pass the entire input data to downstream nodes, and then the downstream nodes retrieve a part of the input data from the shared memory. Instead, we should only pass the part that the downstream node requires, rather than passing the whole data.

In the following example, `a1.inc.bind(i[0])` and `a1.inc.bind(i[1])` read the same mutable object. However, a shared memory channel is not currently supported to be read by the same actor twice, so it will throw a timeout exception.

```python
a1 = Actor.remote(0)
with InputNode() as i:
    branch1 = a1.inc.bind(i[0])
    branch2 = a1.inc.bind(i[1])
    dag = MultiOutputNode([branch1, branch2])

compiled_dag = dag.experimental_compile()

ref = compiled_dag.execute(1, 2)
result = ray.get(ref)
assert result == [1, 3]

compiled_dag.teardown()
```

### Use case

_No response_

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

* Change 1: Remove `class DAGInputAdapter`.
  * Without this PR, the entire input data will be written to the channel, even if a reader only wants to retrieve partial input data via `InputAttributeNode`. Then, the entire input data will be read by the `READ` operation, and the partial input will be retrieved during the `COMPUTE` operation ([code](https://github.com/ray-project/ray/blob/a1be06346e532e656c128ffd2230f06a4d72679e/python/ray/dag/compiled_dag_node.py#L496))
  * In this PR, each `InputAttributeNode` has its own channel, and only the corresponding input data will be written to the channel. Therefore, we no longer need to use `DAGInputAdapter` to retrieve the partial input data during the `COMPUTE` operation.

* Change 2: If the DAG contains any `InputAttributeNode`, create a channel for each `InputAttributeNode`. Then, write the partial input data to the corresponding channel ([code](https://github.com/ray-project/ray/pull/47564/files#diff-5eacb5ce2d9d43e45d6f74bcc6450ba0859c9f853d682a6661cc4237cb614010R425)).

* Change 3: There are some `if/else` statements to handle `InputNode` and `InputAttributeNode` for creating `CachedChannel`. This PR unifies the logic because `InputNode` and different `InputAttributeNode` are no longer considered consumers of only one input channel. Each `InputAttributeNode` has its own channel.

* Change 4: Move `RayDAGArgs` from `compiled_dag_node.py` to `common.py` to avoid importing it inside `_adapt`.
  * Without this, this PR is about 5% slower than the baseline in the case "Benchmark: single actor, no InputAttributeNode". With this change, the performance is almost the same as, or slightly better than, the baseline. See "Benchmark: single actor, no InputAttributeNode" below for more details.

## Benchmark

The baseline is `bb15a35bfd (HEAD -> master, upstream/master) [rllib] remove tf2 tests on pipeline (#47304)`. This PR uses https://github.com/ray-project/ray/pull/47564/commits/ce36f183cc2afdd9c89540c5b32969a644c6a815.

* Experiment setup:
  * 8 actors, 8 InputAttributeNodes
  * payload size: 1 byte (small), 1 MB (medium), 10 MB (large)
  * A Ray cluster with 1 node

* Some factors may influence performance
  * Number of `InputAttributeNode`
    * In the baseline, `8 (InputAttributeNode) * payload_size` will be written to a single `MutableObject`. In this PR, 8 pieces of data, each with a size of `payload_size`, will be written to 8 different `MutableObjects`. I guess the time it takes to write data into shared memory will be shorter in the baseline compared to this PR. If the number of `InputAttributeNode` increases, the difference will be larger.
  * `payload_size` & number of consumer
    * In the baseline, each consumer needs to deserialize the entire input data (`8 (InputAttributeNode) * payload_size`). In this benchmark, each actor needs to deserialize `8 (InputAttributeNode) * payload_size`, and there are 8 actors. In this PR, each actor only needs to deserialize `payload_size`. Hence, if `payload_size` increases, this PR will be better than the baseline.
    * The `payload_size` may also determine whether to trigger the resize of the mutable object.
  * Number of Ray nodes
    * In the baseline, if consumers are on multiple Ray nodes, the entire input data, with a size of `8 (InputAttributeNode) * payload_size`, will be copied to all these Ray nodes via gRPC. In this PR, only the `InputAttributeNode` used by the consumers will be copied to the node where the consumer is located.

* Experiment result
  * small payload (1 byte)
    * baseline: 3725.18
    * this PR: 4625.1  
  * medium payload (1 MB)
    * baseline: 71.24
    * this PR: 218.98
  * large payload (10 MB)
    * baseline: 4.17
    * this PR: 9.6

* With this PR
  Screenshot 2024-09-14 at 6 34 09 AM

* Baseline
  Screenshot 2024-09-14 at 6 34 22 AM

### Benchmark: multiple node

Run [this script](https://gist.github.com/kevin85421/a635886d53cdec8d544620ba09070068).

* Experiment setup
  * 4 Ray nodes
  * 4 actors, 4 InputAttributeNode
  * payload_size = 10 MB 
* Experiment result (the time it takes to execute 5 times.)
  * This PR: 13.63s
  * Baseline: 34.98s

* With this PR:
  Screenshot 2024-09-14 at 7 07 54 AM
* Baseline:
  image

### Benchmark: single actor, no InputAttributeNode

https://github.com/ray-project/ray/blob/d07975e1544e7061676c69f1563cbc0894941f5c/python/ray/_private/ray_experimental_perf.py#L163-L194

* Commit under test: https://github.com/ray-project/ray/pull/47564/commits/bbd501d9514d673de36b6a447c2f3668c1a5ab92
  * This is a different commit from the one used in the benchmarks above. This commit avoids importing `RayDAGArgs` inside `_adapt`. Without this commit, this PR is about 5% slower than the baseline. With this commit, the performance is almost the same as, or slightly better than, the baseline.
* Baseline: `4b2f6a07ca (HEAD -> master, upstream/master) [Core][aDAG] support multi readers in multi node when dag is created from an actor (#47601)`

* With this PR
  Screenshot 2024-09-14 at 2 46 52 PM

* Without this PR
  Screenshot 2024-09-14 at 2 46 41 PM

### Benchmark: single actor takes all args.

* ***I expected this PR to be slower than the baseline, but it is faster. It's interesting. If we only call execute a few times (e.g., 100), the baseline is faster than this PR. However, if we continuously call execute more times, this PR becomes faster than the baseline.***
  * Try [this gist](https://gist.github.com/kevin85421/e712d113b60289499e8073474cb7c96c) to execute the DAG 100 times and 1000 times.
    * 100 times:
      * Baseline:  `Average time for a single compiled DAG call: 0.0002106591360643506`.
      * This PR: `Average time for a single compiled DAG call: 0.0002376893116161227`
    * 1000 times:
      * Baseline: `Average time for a single compiled DAG call: 0.00020164441270753743`
      * This PR: `Average time for a single compiled DAG call: 0.00018878400512039662`

```python
actor = DAGActor.remote()
n_args = 8
with InputNode() as inp:
    dag = actor.echo_multiple.bind(*[inp[i] for i in range(n_args)])
payload_size = 1
results += timeit(
    "[unstable] single-actor with all args with small payloads DAG calls, "
    "n=1 actors",
    lambda: ray.get(dag.execute(*[b"x" * payload_size for _ in range(n_args)])),
)
compiled_dag = dag.experimental_compile()
results += timeit(
    "[unstable] single-actor with all args with small payloads DAG calls, "
    "n=1 actors",
    lambda: _exec(compiled_dag, num_args=n_args, payload_size=payload_size),
)
compiled_dag.teardown()
```

* Commit under test: https://github.com/ray-project/ray/pull/47564/commits/bbd501d9514d673de36b6a447c2f3668c1a5ab92
* Baseline: `4b2f6a07ca (HEAD -> master, upstream/master) [Core][aDAG] support multi readers in multi node when dag is created from an actor (#47601)`

* With this PR
  Screenshot 2024-09-14 at 4 16 27 PM

* Baseline
  Screenshot 2024-09-14 at 4 16 48 PM

## Related issue number

Closes #47366

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
lmk when you want me to take a look!

File: python/ray/experimental/channel/common.py
Comment (kevin85421) [inline python/ray/experimental/channel/common.py:522]: TODO

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (kevin85421) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:768]: The tests above are just indented. Below are the new tests.

File: python/ray/experimental/channel/common.py
Comment (kevin85421) [inline python/ray/experimental/channel/common.py:377]: TODO

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:243]: Change 1

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:1202]: Change 2

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:1298]: Change 3

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1206]:
```suggestion
                input_node_to_reader_and_node_set: Dict[
```

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1239]: not technically related to this PR, but why do we have separate list for output_idxs and output_channels? maybe we should combine it to a single dict

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (rkooo567) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:770]: QQ: is it a behvior change or the same?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (rkooo567) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:878]: can you do multi args + torch tensor type too?

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1247]: can you give me more details about what this is doing? I am a little confused because the task has output channel, but then we have separate input attr task, and they have its own output channels? why is that?

**Review by rkooo567 — COMMENTED** [review]
> In this PR, each InputAttributeNode has its own channel, and only the corresponding input data will be written to the channel. Therefore, we no longer need to use DAGInputAdapter to retrieve the partial input data during the COMPUTE operation.

Great! I think this means we can avoid unnecessary deserialization when you have sharded inputs

Can you also check if this test case works with this PR?
```python3
 with ray.dag.InputNode() as dag_input:
        dag_input.with_type_hint(TorchTensorType())
        dag_outs = []
        for microbatch_idx in range(NUM_MICROBATCHES):
            microbatch = dag_input[microbatch_idx]
            stage_fwd_out = s1.execute_forward.bind(microbatch)
            stage_fwd_out = stage_fwd_out.with_type_hint(TorchTensorType(transport="nccl"))
            stage_fwd_out = s2.execute_forward.bind(stage_fwd_out)
            dag_outs.append(stage_fwd_out)
        
        grad_out = dag_input[NUM_MICROBATCHES]
        for _ in range(NUM_MICROBATCHES):
            stage_bwd_out = s2.execute_backward.bind(grad_out)
            stage_bwd_out = stage_bwd_out.with_type_hint(TorchTensorType(transport="nccl"))
            stage_bwd_out = s1.execute_backward.bind(stage_bwd_out)
            dag_outs.append(stage_bwd_out)
        
        dag = ray.dag.MultiOutputNode(dag_outs)
```

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:770]: No, it just hasn't been tested, and some ADAG logic is implemented based on this assumption. The check is in `preprocess`.

https://github.com/ray-project/ray/blob/35b453445b49ad286c08099953c000c3942acf6a/python/ray/dag/compiled_dag_node.py#L974-L978

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1247]: 1. This ensures consistent behavior with `ClassMethodNode` when `is_class_method_output` is set to True.

https://github.com/ray-project/ray/blob/35b453445b49ad286c08099953c000c3942acf6a/python/ray/dag/compiled_dag_node.py#L1241-L1253

2. When creating a `CachedChannel`, we need to handle `InputAttributeNode` separately because its output channel is empty, which will cause the assertion `assert len(upstream_task.output_channels) == 1` to fail.

https://github.com/ray-project/ray/blob/35b453445b49ad286c08099953c000c3942acf6a/python/ray/dag/compiled_dag_node.py#L1343-L1356

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1239]: I don't know, but I think using a list of tuples or a dictionary is better. Maybe we can revisit this when we have a more granular execution schedule generation mechanism. To generate a granular execution schedule, we should be able to control the order of `output_channels` in a writer.

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Update: https://github.com/ray-project/ray/pull/47564/commits/3c8cf02954f3d112a77076dfdda5c31ef9fcdfbd

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:903]: Added https://github.com/ray-project/ray/pull/47564/commits/3c8cf02954f3d112a77076dfdda5c31ef9fcdfbd

**Comment by kevin85421** [comment]
> Can you also check if this test case works with this PR?

Added https://github.com/ray-project/ray/pull/47564/commits/d17a9d4ee9c3de63bd853aec4ccb0bf314e4ff7b. Let's wait and see if I configured the CI correctly.

File: python/ray/dag/compiled_dag_node.py
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1300]: So this always creates multiple input channels? Wondering what would be the performance implications (e.g., locking on multiple channels), maybe we need some evaluations.

Alternatively, can we use one channel if it's not absolutely necessary to create multiple channels?

**Review by ruisearch42 — COMMENTED** [review]
Made a quick pass. The approach looks reasonable.

**Comment by kevin85421** [comment]
I faced the same issue as described in https://github.com/ray-project/ray/issues/47093 when I added the GPU tests to the CI. I can reproduce the issue on Anyscale workspace, and find a solution to solve it. See #47093 for more details. However, the solution needs to install a new dependency and downgrade an existing dependency, so I will do it in a separate PR.

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:1300]: @kevin85421 can you run a benchmark with 20 arguments before/after this PR?

```
dag.execute(a, b, c, d, e....) # 20 arguments
```

I think if there's a regression, what we can do is to group inputs that are consumed together to a single channel. 

For example;

```
with InputNode() as inp:
    dag = a.f.bind(inp[0], inp[1])
```

If you have this kind of DAG, we know both inp[0] and inp[1] is consumed by only one downstream consumer, so we can put them to a single channel to reduce channel read overhead

**Review by rkooo567 — APPROVED** [review]
LGTM. 

1. Plz follow up with GPU CI!
2. Let's run a benchmark. I don't think we need to block this PR though.

File: python/ray/dag/compiled_dag_node.py
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:1300]: > If the InputNode is consumed directly (i.e., there is no InputAttributeNode), only one input channel will be created. If there are any InputAttributeNodes, allocate a channel for each InputAttributeNode.

Previously only one channel is created even if there are multiple `InputAttributeNode`s. I think we should understand if there is any performance degradation (although not necessarily a blocker for this PR), and if so, fall back to the original approach if possible.

> can you run a benchmark with 20 arguments before/after this PR?

We can add this to the microbenchmark.

Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:1267]: > Previously only one channel is created even if there are multiple InputAttributeNodes. I think we should understand if there is any performance degradation (although not necessarily a blocker for this PR), and if so, fall back to the original approach if possible.

Btw, theoretically this should have better performance. The original approach has do deserialize the entire object although we only use a part of them.

```
with InputNode() as inp:
    a.f.bind(inp[0])
    a.f.bind(inp[0])
    a.f.bind(inp[0])
```

**Comment by rkooo567** [comment]
@kevin85421 can you also share the benchmark result before/after?

**Comment by kevin85421** [comment]
@rkooo567 I have already added the benchmark results in the PR description. I will add the case that doesn't have `InputAttributeNode` tmr.

File: python/ray/_private/ray_experimental_perf.py
Comment (rkooo567) [inline python/ray/_private/ray_experimental_perf.py:308]: can you add a case where all input attribute args go to a single actor?

```
actor = DAGActor.remote()
with InputNode() as inp:
    dag = actor.echo.bind(*inp)

adag = dag.experimental_compile()
adag.execute()
```

Comment (rkooo567) [reply python/ray/_private/ray_experimental_perf.py:308]: I think this is the case that may have regression.

**Review by rkooo567 — COMMENTED** [review]
okay, great. I think shard case is expected to have higher performance as we expected. I think we should also add test case for N -> 1 case (I added a relevant comment)

**Comment by rkooo567** [comment]
also before merging, let's run this in release pipeline so that we just make sure the microbenchmark_unstable doesn't crash

File: python/ray/dag/compiled_dag_node.py
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1259]: Added several benchmarks to CI, and added more details in the PR description.

File: python/ray/_private/ray_experimental_perf.py
Comment (kevin85421) [reply python/ray/_private/ray_experimental_perf.py:311]: See "Benchmark: single actor takes all args." in the PR description. It is faster than the baseline. Weird.

File: python/ray/experimental/channel/common.py
Comment (kevin85421) [inline python/ray/experimental/channel/common.py:22]: Change 4: Move `RayDAGArgs` from `compiled_dag_node.py` to `common.py` to avoid importing it inside `_adapt`.

**Comment by kevin85421** [comment]
The microbenchmark release test fails ([Buildkite run](https://buildkite.com/ray-project/release/builds/22651)). The following screenshot shows the error message. However, it is not related to this PR. The master branch also fails with the same error message ([example](https://buildkite.com/ray-project/release/builds/22636)).

image

File: python/ray/_private/ray_experimental_perf.py
Comment (rkooo567) [inline python/ray/_private/ray_experimental_perf.py:329]: can you do more extreme case, something like 128?

**Review by rkooo567 — APPROVED** [review]
nice. makes sense multi node we are way better.

To decide if we need to auto-grouping (grouping inputs based on downstream consmers) input channels, I want to do more extreme benchmark. I will merge the PR, but plz follow up with the result in the command. 

Can you check running 128 inputs with a single actor and see if it matches perf parity?

**Pull request merged**
